keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,501 @@
+- The types and their behaviors are easier to understand since they are
+described together.
+- They are less error-prone, and discourage people from using anti-patterns
+like "cross-field dependency defaults" (default depends on another
+field), non-deterministic or complicated defaults. These patterns
+might be useful in limited circumstances, but it should be easier to do
+the right thing in most cases.
+- Removes the inconsistency with CRDs that already support this
+mechanism.
+- Solves bug where associative keys needs to either be required or
+defaulted. Default keys will now be able to be specified in the OpenAPI
+for built-in types, and SSA will be able to fill in the missing keys.
+
+### Goals
+
+The goal is to add a new `// +default` marker to our current built-in Go
+IDL. That marker will be transformed into the OpenAPI `default` field
+and then routed to defaulting functions so that defaulting can be done
+declaratively. This new marker would also be introduced in kubebuilder,
+and we will make the behavior equivalent between built-in types and CRDs
+so that built-in types can be directly used in kubebuilder, and the
+declarative defaults will be automatically carried along.
+
+### Non-Goals
+
+While this proposal is probably paving the way for additional OpenAPI
+based validation, validation is entirely out-of-scope for this project.
+
+This proposal also doesn't intend to remove the existing defaulting
+mechanism since some defaulting still requires non-static values that
+need to be computed and can't be defined declaratively. We could replace
+some defaulting that doesn't require to be done as code though.
+
+## Proposal
+
+It is important to us that the OpenAPI definition of a type should
+result in the same behavior, whether the type is built-in or a CRD.
+
+In order to do that, we need to change how CRDs are defaulted, and
+create a defaulting mechanism for built-in that would behave the same
+way.
+
+### CRD Normalization
+
+Null values in CRDs are currently not defaulted. On the other hand,
+`null` typically means "unspecified" in json/yaml. JSon Merge patch
+interprets a "null" value for a field as a mean to delete it.
+
+On the other hand, while `{field: null}` and `{}` probably have the same
+meaning, they do not "look" the same and are hard to compare for
+clients. They are also currently treated differently by Kubernetes
+for defaulting purposes.
+
+We propose the creation of a new top-level openapi flag:
+`x-kubernetes-remove-nulls`. This flag will be used as a ratcheting
+mechanism as we migrate to this new semantic throughout all Kubernetes
+types (which would be how all types eventually work). When present and
+set to true, the apiserver will use new semantics and remove all null
+values from CR objects. When false or missing, the existing and
+deprecated semantics will be kept. Alternatively, this flag could be set
+as a field of the CRD since this process will only apply to CRDs.
```

Comment:
This mechanism is quite similar to [pruning](https://kubernetes.io/blog/2019/06/20/crd-structural-schema/#pruning-don-t-preserve-unknown-fields) which was used for ratcheting as well. Since pruning is enabled by the `preserveUnknownFields` CRD field, I'm currently in favor of adding a similar flag for null removal, for consistency. E.g. `preserveNullValues`, with a plan to eventually make it default to false like was done for `preserveUnknownFields` when CRDs went to v1. cc @sttts

Reply:
Also, I suspect we would run the pruning of null at the exact same time and location in code, since we already have mechanisms in place to recursively descend the objects and apply these forms of normalization. That's a 5 liner there. thanks!